docs/architecture/adr-064-batch-verification.md

```diff
@@ -0,0 +1,85 @@
+
+## Decision
+
+Adopt Batch Verification.
+
+## Detailed Design
+
+A new interface will be introduced. This interface will have three methods `NewBatchVerifier`, `Add` and `VerifyBatch`.
+
+```go
+type BatchVerifier interface {
+  NewBatchVerifier() BatchVerifier // NewBatchVerifier create a new verifier where keys, signatures and messages can be added as entries
+  Add(key crypto.Pubkey, signature, message []byte) error // Add appends an entry into the BatchVerifier.
+  Verify() bool // Verify verifies all the entries in the BatchVerifier. If the verification fails it is unknown which entry failed and each entry will need to be verified individually.
+}
+```
+
+- `NewBatchVerifier` creates a new verifier. This verifier will be populated with entries to be verified. 
+- `Add` adds an entry to the Verifier. Add accepts a public key and two slice of bytes (signature and message). 
+- `Verify` verifies all the entires. At the end of Verify if the underlying API does not reset the Verifier to its initial state (empty), it should be done here. This prevents accidentally reusing the verifier with entries from a previous verification.
+
+Above there is mention of an entry. An entry can be constructed in many ways depending on the needs of the underlying curve. A simple approach would be:
+
+```go
+type entry struct {
+  pubKey crypto.Pubkey
+  signature []byte
+  message []byte
+}
+```
+
+The main reason this approach is being taken is to prevent simple mistakes. Some APIs allow the user to create three slices and pass them to the `VerifyBatch` function but this relies on the user to safely generate all the slices (see example below). We would like to minimize the possibility of making a mistake.
+
+```go
+func Verify(keys []crypto.Pubkey, signatures, messages[][]byte) bool
+```
+
+This change will not affect any users in anyway other than faster verification times.
+
+This new api will be used for verification in both consensus and block syncing. Within the current Verify functions there will be a check to see if the key types supports the BatchVerification API. If it does it will execute batch verification, if not single signature verification will be used. 
+
+#### Consensus
+
+  The process within consensus will be to wait for 2/3+ of the votes to be received, once they are received `Verify()` will be called to batch verify all the messages. The messages that come in after 2/3+ has been verified will be individually verified. 
```

Comment:
Can signatures post 2/3+ be verified optionally?

Reply:
Yes, this would be a local configuration instead of a network wide change. The `timeout-prevote` would be set to zero.